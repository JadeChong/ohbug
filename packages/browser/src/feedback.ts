import { createEvent, collector } from '@ohbug/core'
import { getGlobal } from '@ohbug/utils'

export const prefix = 'ohbug-feedback'

const styles = `
  .${prefix}-root {
    display: block;
    width: 100%;
    height: 100%;
    position: fixed;
    left: 0;
    top: 0;
    z-index: 999;
    background: rgba(0, 0, 0, 0.5);
    overflow: auto;
  }
  .${prefix}-root * {
    box-sizing: border-box;
    color: #666;
  }

  .${prefix}-dialog {
    max-width: 800px;
    margin: 5% auto 0;
    padding: 36px;
    background: white;
    border-radius: 5px;
  }

  .${prefix}-dialog header {
    text-align: center;
    border-bottom: 1px solid #eee;
    padding-bottom: 18px;
  }
  .${prefix}-dialog header p {
    color: #aaa;
  }

  .${prefix}-content .${prefix}-field {
    overflow: hidden;
  }
  .${prefix}-content .${prefix}-field label {
    display: block;
    margin: 18px 0 10px;
    font-size: 16px;
  }
  .${prefix}-content .${prefix}-field input, .${prefix}-content .${prefix}-field textarea {
    width: 100%;
    padding: 8px 12px;
    background: white;
    outline: none;
    border: 1px solid #ddd;
    border-radius: 2px;
  }
  .${prefix}-submit {
    margin-top: 18px;
  }
  .${prefix}-submit .${prefix}-submit-btn {
    outline: none;
    border: none;
    padding: 0 20px;
    height: 36px;
    background-color: #3751ff;
    color: white;
    font-size: 16px;
    border-radius: 4px;
    cursor: pointer;
    box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
    transition: background 0.3s ease;
    margin-right: 12px;
  }
  .${prefix}-submit .${prefix}-submit-btn:hover {
    background-color: #617bff;
  }
  .${prefix}-submit-close {
    cursor: pointer;
  }
  .${prefix}-submit-close:hover {
    text-decoration: underline;
  }
`

interface OElement {
  type: string
  props?: any
  children?: OElement | OElement[] | string | number
}
function render(element: OElement, container: Element): HTMLElement {
  const { type, props, children } = element

  const window = getGlobal<Window>()

  const dom = window.document.createElement(type)

  // map props
  for (const propName in props) {
    const value = props[propName]
    if (propName === 'style') {
      const { style } = props
      for (const styleKey in style) {
        // @ts-ignore
        dom[propName][styleKey] = style[styleKey]
      }
      continue
    } else if (propName[0] === 'o' && propName[1] === 'n') {
      const name = propName.slice(2).toLowerCase()
      dom.addEventListener(name, value)
    } else if (!value) {
      dom.removeAttribute(propName)
    } else {
      dom.setAttribute(propName, value)
    }
  }

  // handle children
  if (children) {
    if (typeof children === 'string' || typeof children === 'number') {
      const text = window.document.createTextNode(children.toString())
      dom.appendChild(text)
    } else if (Array.isArray(children)) {
      children.forEach(child => {
        render(child, dom)
      })
    } else {
      render(children, dom)
    }
  }

  // append
  container.appendChild(dom)

  return dom
}

function feedback(): HTMLElement {
  let dom: HTMLElement
  const handleClose = () => {
    window.document.body.removeChild(dom)
  }
  const element: OElement = {
    type: 'div',
    props: {
      class: `${prefix}-root`,
      'data-testid': `${prefix}-root`
    },
    children: [
      {
        type: 'style',
        children: styles
      },
      {
        type: 'div',
        props: {
          class: `${prefix}-dialog`
        },
        children: [
          {
            type: 'header',
            children: [
              {
                type: 'h2',
                children: `似乎发生了一些问题。`
              },
              {
                type: 'p',
                children: `如果您愿意提供帮助，可以在下方填写问题以便告诉我们发生了什么。`
              }
            ]
          },
          {
            type: 'form',
            props: {
              action: '',
              method: 'post',
              onSubmit: (e: Event): void => {
                e.preventDefault()

                const name = (window.document.querySelector(`#${prefix}-name`) as HTMLInputElement)
                  .value
                const email = (window.document.querySelector(
                  `#${prefix}-email`
                ) as HTMLInputElement).value
                const comments = (window.document.querySelector(
                  `#${prefix}-comments`
                ) as HTMLTextAreaElement).value
                const data = {
                  name,
                  email,
                  comments
                }
                const event = createEvent('FEEDBACK', data, 'feedback')
                collector(event, 'sync')

                handleClose()
              }
            },
            children: [
              {
                type: 'div',
                props: {
                  class: `${prefix}-content`
                },
                children: [
                  {
                    type: 'div',
                    props: {
                      class: `${prefix}-field`
                    },
                    children: [
                      {
                        type: 'label',
                        props: {
                          for: `${prefix}-name`
                        },
                        children: '姓名'
                      },
                      {
                        type: 'input',
                        props: {
                          type: 'text',
                          id: `${prefix}-name`,
                          placeholder: 'Cxy',
                          maxlength: 128,
                          required: true
                        }
                      }
                    ]
                  },
                  {
                    type: 'div',
                    props: {
                      class: `${prefix}-field`
                    },
                    children: [
                      {
                        type: 'label',
                        props: {
                          for: `${prefix}-email`
                        },
                        children: '电子邮件'
                      },
                      {
                        type: 'input',
                        props: {
                          type: 'email',
                          id: `${prefix}-email`,
                          placeholder: 'Cxy@example.com',
                          maxlength: 75,
                          required: true
                        }
                      }
                    ]
                  },
                  {
                    type: 'div',
                    props: {
                      class: `${prefix}-field`
                    },
                    children: [
                      {
                        type: 'label',
                        props: {
                          for: `${prefix}-comments`
                        },
                        children: '发生了什么'
                      },
                      {
                        type: 'textarea',
                        props: {
                          id: `${prefix}-comments`,
                          placeholder: '我点击了按钮 A',
                          required: true,
                          rows: 10,
                          cols: 40
                        }
                      }
                    ]
                  }
                ]
              },
              {
                type: 'div',
                props: {
                  class: `${prefix}-submit`
                },
                children: [
                  {
                    type: 'button',
                    props: {
                      type: 'submit',
                      class: `${prefix}-submit-btn`
                    },
                    children: '提交'
                  },
                  {
                    type: 'a',
                    props: {
                      class: `${prefix}-submit-close`,
                      onClick: () => {
                        handleClose()
                      }
                    },
                    children: '关闭'
                  }
                ]
              }
            ]
          }
        ]
      }
    ]
  }
  dom = render(element, window.document.body)
  return dom
}

export default feedback
