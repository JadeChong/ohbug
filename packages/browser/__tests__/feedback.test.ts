import initBrowser from '../src/init'
import feedback from '../src/feedback'

const apiKey = 'test_id'
const config = { apiKey }

describe('browser feedback', () => {
  beforeAll(() => {
    initBrowser(config)
  })

  it('should render feedback', () => {
    const dom = feedback()
    expect(dom).toMatchSnapshot()
  })
})
