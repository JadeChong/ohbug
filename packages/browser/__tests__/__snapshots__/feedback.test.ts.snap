// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`browser feedback should render feedback 1`] = `
<div
  class="ohbug-feedback-root"
  data-testid="ohbug-feedback-root"
>
  <style>
    
  .ohbug-feedback-root {
    display: block;
    width: 100%;
    height: 100%;
    position: fixed;
    left: 0;
    top: 0;
    z-index: 999;
    background: rgba(0, 0, 0, 0.5);
    overflow: auto;
  }
  .ohbug-feedback-root * {
    box-sizing: border-box;
    color: #666;
  }

  .ohbug-feedback-dialog {
    max-width: 800px;
    margin: 5% auto 0;
    padding: 36px;
    background: white;
    border-radius: 5px;
  }

  .ohbug-feedback-dialog header {
    text-align: center;
    border-bottom: 1px solid #eee;
    padding-bottom: 18px;
  }
  .ohbug-feedback-dialog header p {
    color: #aaa;
  }

  .ohbug-feedback-content .ohbug-feedback-field {
    overflow: hidden;
  }
  .ohbug-feedback-content .ohbug-feedback-field label {
    display: block;
    margin: 18px 0 10px;
    font-size: 16px;
  }
  .ohbug-feedback-content .ohbug-feedback-field input, .ohbug-feedback-content .ohbug-feedback-field textarea {
    width: 100%;
    padding: 8px 12px;
    background: white;
    outline: none;
    border: 1px solid #ddd;
    border-radius: 2px;
  }
  .ohbug-feedback-submit {
    margin-top: 18px;
  }
  .ohbug-feedback-submit .ohbug-feedback-submit-btn {
    outline: none;
    border: none;
    padding: 0 20px;
    height: 36px;
    background-color: #3751ff;
    color: white;
    font-size: 16px;
    border-radius: 4px;
    cursor: pointer;
    box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
    transition: background 0.3s ease;
    margin-right: 12px;
  }
  .ohbug-feedback-submit .ohbug-feedback-submit-btn:hover {
    background-color: #617bff;
  }
  .ohbug-feedback-submit-close {
    cursor: pointer;
  }
  .ohbug-feedback-submit-close:hover {
    text-decoration: underline;
  }

  </style>
  <div
    class="ohbug-feedback-dialog"
  >
    <header>
      <h2>
        似乎发生了一些问题。
      </h2>
      <p>
        如果您愿意提供帮助，可以在下方填写问题以便告诉我们发生了什么。
      </p>
    </header>
    <form
      method="post"
    >
      <div
        class="ohbug-feedback-content"
      >
        <div
          class="ohbug-feedback-field"
        >
          <label
            for="ohbug-feedback-name"
          >
            姓名
          </label>
          <input
            id="ohbug-feedback-name"
            maxlength="128"
            placeholder="Cxy"
            required="true"
            type="text"
          />
        </div>
        <div
          class="ohbug-feedback-field"
        >
          <label
            for="ohbug-feedback-email"
          >
            电子邮件
          </label>
          <input
            id="ohbug-feedback-email"
            maxlength="75"
            placeholder="Cxy@example.com"
            required="true"
            type="email"
          />
        </div>
        <div
          class="ohbug-feedback-field"
        >
          <label
            for="ohbug-feedback-comments"
          >
            发生了什么
          </label>
          <textarea
            cols="40"
            id="ohbug-feedback-comments"
            placeholder="我点击了按钮 A"
            required="true"
            rows="10"
          />
        </div>
      </div>
      <div
        class="ohbug-feedback-submit"
      >
        <button
          class="ohbug-feedback-submit-btn"
          type="submit"
        >
          提交
        </button>
        <a
          class="ohbug-feedback-submit-close"
        >
          关闭
        </a>
      </div>
    </form>
  </div>
</div>
`;
